File: firestarter/src/main/java/com/googlecode/blaisemath/firestarter/editor/MPanelEditorSupport.java
package com.googlecode.blaisemath.firestarter.editor;

/*
 * #%L
 * Firestarter
 * --
 * Copyright (C) 2009 - 2025 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.awt.Component;
import javax.swing.JPanel;

/**
 * Generic super class for custom editors that use a panel for layout and display.
 *
 * @author Elisha Peterson
 */
public abstract class MPanelEditorSupport extends MPropertyEditorSupport {
    
    protected JPanel panel;

    public MPanelEditorSupport() {
    }

    /** Initializes the customized component... this is not called initially in
     * case subclasses do not need the custom panel.
     */
    protected abstract void initCustomizer();

    @Override
    public boolean supportsCustomEditor() {
        return true;
    }

    @Override
    public Component getCustomEditor() {
        if (panel == null) {
            initCustomizer();
            initEditorValue();
        }
        panel.addPropertyChangeListener("enabled", evt -> {
            for (int i = 0; i < panel.getComponentCount(); i++) {
                panel.getComponent(i).setEnabled(panel.isEnabled());
            }
        });
        return panel;
    }
}

File: firestarter/src/main/java/com/googlecode/blaisemath/firestarter/editor/MPanelEditorSupport.kt
package com.googlecode.blaisemath.firestarter.editor

import com.googlecode.blaisemath.app.ApplicationMenuConfig
import com.googlecode.blaisemath.firestarter.editor.EditorRegistrationTest
import com.googlecode.blaisemath.firestarter.property.BeanTreeNode
import com.googlecode.blaisemath.firestarter.property.CustomBean
import com.googlecode.blaisemath.firestarter.property.HelloWorldTestFrame
import com.googlecode.blaisemath.firestarter.property.IndexedBean
import com.googlecode.blaisemath.firestarter.property.IndexedBean.Indexed2
import com.googlecode.blaisemath.firestarter.property.IndexedBean.TestEnum
import com.googlecode.blaisemath.firestarter.property.NumberBean
import com.googlecode.blaisemath.firestarter.property.PointBean
import com.googlecode.blaisemath.firestarter.property.PropertySheetDialogTestFrame
import com.googlecode.blaisemath.firestarter.property.ShapeBean
import com.googlecode.blaisemath.firestarter.property.TestPropertyModel
import com.googlecode.blaisemath.firestarter.property.Testing
import com.googlecode.blaisemath.ui.PropertyActionPanel
import junit.framework.TestCase
import org.apache.commons.math.stat.descriptive.SummaryStatistics
import org.junit.Before
import org.junit.BeforeClassimport
import java.beans.PropertyChangeEvent
import java.beans.PropertyChangeListener
import javax.swing.JPanel

java.awt.*
/*
 * #%L
 * Firestarter
 * --
 * Copyright (C) 2009 - 2019 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */ /**
 * Generic super class for custom editors that use a panel for layout and display.
 *
 * @author Elisha Peterson
 */
abstract class MPanelEditorSupport : MPropertyEditorSupport() {
    protected var panel: JPanel? = null

    /** Initializes the customized component... this is not called initially in
     * case subclasses do not need the custom panel.
     */
    protected abstract fun initCustomizer()
    override fun supportsCustomEditor(): Boolean {
        return true
    }

    override fun getCustomEditor(): Component? {
        if (panel == null) {
            initCustomizer()
            initEditorValue()
        }
        panel.addPropertyChangeListener("enabled", PropertyChangeListener { evt: PropertyChangeEvent? ->
            for (i in 0 until panel.getComponentCount()) {
                panel.getComponent(i).isEnabled = panel.isEnabled()
            }
        })
        return panel
    }
}